src/lib/support/tests/TestStringCircularBuffer.cpp

```diff
@@ -0,0 +1,143 @@
+
+    NL_TEST_ASSERT(inSuite, buffer.Pop() == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, buffer.IsEmpty());
+}
+
+void TestPushInvalid(nlTestSuite * inSuite, void * inContext)
+{
+    uint8_t storage[10];
+    StringCircularBuffer buffer(storage, sizeof(storage));
+    const unsigned char s[] = { '1', '2', '3', '4', '5', '6', '7', '8', '9', '0' };
+    NL_TEST_ASSERT(inSuite, buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s, 8)) == CHIP_ERROR_INVALID_ARGUMENT);
+    NL_TEST_ASSERT(inSuite, buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s, 9)) == CHIP_ERROR_INVALID_ARGUMENT);
+    NL_TEST_ASSERT(inSuite, buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s, 7)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+}
+
+void TestPushWrap(nlTestSuite * inSuite, void * inContext)
+{
+    uint8_t storage[10];
+    const unsigned char s1[] = { '1', '2', '3', '4', '5', '6', '7', '8', '9', '0' };
+    const unsigned char s2[] = { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j' };
+    unsigned char o[sizeof(s1)];
+
+    // Total space = 9
+    StringCircularBuffer buffer(storage, sizeof(storage));
+    NL_TEST_ASSERT(inSuite, buffer.IsEmpty());
+
+    // Used = 6, Free = 3
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s1, 4)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 4);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(o) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s1, o, 4) == 0);
+
+    // Used = 6, Free = 3, s1 was discarded due to lack of storage
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s2, 4)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 4);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(o) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s2, o, 4) == 0);
+
+    // Used = 9, Free = 0
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s1, 1)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 4);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(o) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s2, o, 4) == 0);
+
+    // Used = 3, Free = 6
+    NL_TEST_ASSERT(inSuite, buffer.Pop() == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 1);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(o) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s1, o, 1) == 0);
+}
+
+int Setup(void * inContext)
+{
+    return SUCCESS;
+}
+
+int Teardown(void * inContext)
+{
+    return SUCCESS;
+}
+
+} // namespace
+
+#define NL_TEST_DEF_FN(fn) NL_TEST_DEF("Test " #fn, fn)
+/**
+ *   Test Suite. It lists all the test functions.
+ */
+static const nlTest sTests[] = { NL_TEST_DEF_FN(TestPushPop), NL_TEST_DEF_FN(TestPushInvalid), NL_TEST_DEF_FN(TestPushWrap),
```

Comment:
Please add tests at least for the following:

1. What happens if capacity < 3.
2. What happens with incoming ByteSpans with malicious lengths like `static_cast<size_t>(-1)`
3. What happens when trying to push something that would, when added to what's there already, exactly equal mCapacity in total size.
4. Explicitly exercise the case when mDataEnd would end up equaling mCapacity (and should wrap to 0).
5. Exercise the case when mDataStart ends up equal to mCapacity (and should wrap to 0).  That is, a case when `mDataEnd` became 0, then we pushed something, then we popped all the things that used to be there.
6. Any other interesting edge cases you think of.